### PR TITLE
Fix `unused_parens` triggers on macro by example code

### LINKED ISSUE #66295: `unused_parens` triggers on macro by example code.
The latest beta contains a new? l`unused_parens` lint that triggers on macro by example code. This does not necessarily make sense because macros could be used to construct nearly arbitrary custom syntax for example something requiring to wrap a single element into a parenthesis. 
The strange thing here is that seems to only happen on beta, not on nightly, not on current stable.

Found via [the diesel CI](https://dev.azure.com/diesel-rs/diesel/_build/results?buildId=618)

Error message:

```
    --> diesel/src/macros/mod.rs:274:27
     |
274  |               primary_key = (id),
     |                             ^^^^ help: remove these parentheses
…     |                             ^^^^ help: remove these parentheses
...
1085 | /     table! {
1086 | |         use sql_types::*;
1087 | |         use macros::tests::my_types::*;
1088 | |
...    |
1092 | |         }
1093 | |     }
     | |_____- in this macro invocation
```

rustc version: `rustc 1.40.0-beta.1 (76b40532a 2019-11-05)`

### PR DESCRIPTION
**Pull request by weiznich**
Fix #66295

Unfortunately this does also break [an existing test](https://github.com/rust-lang/rust/blob/4787e97475de6be9487e3d9255a9c2d3c0bf9252/src/test/ui/lint/issue-47775-nested-macro-unnecessary-parens-arg.rs#L22). I'm not sure how to handle that, because that seems to be quite similar to the allowed cases

If this gets accepted it would be great to backport this fix to beta.

### CODE REVIEW

File: src/librustc_lint/unused.rs [hunk 1]
```diff
@@ -355,7 +355,11 @@ impl UnusedParens {
         match value.kind {
             ast::ExprKind::Paren(ref inner) => {
                 if !Self::is_expr_parens_necessary(inner, followed_by_block) &&
-                    value.attrs.is_empty() {
+                    value.attrs.is_empty() &&
+                    !MultiSpan::from(value.span).primary_span()
+                        .map(|span| span.from_expansion())
```
Comment (lzutao) [inline src/librustc_lint/unused.rs]:
```suggestion
                        .map_or(false, |span| span.from_expansion())
```

File: src/librustc_lint/unused.rs [hunk 2]
```diff
@@ -355,7 +355,11 @@ impl UnusedParens {
...
+                        .unwrap_or(false)
```
Comment (lzutao) [inline src/librustc_lint/unused.rs]:
```suggestion
```

File: src/librustc_lint/unused.rs [hunk 3]
```diff
@@ -355,7 +355,10 @@ impl UnusedParens {
         match value.kind {
             ast::ExprKind::Paren(ref inner) => {
                 if !Self::is_expr_parens_necessary(inner, followed_by_block) &&
-                    value.attrs.is_empty() {
+                    value.attrs.is_empty() &&
+                    !MultiSpan::from(value.span).primary_span()
+                        .map_or(false, |span| span.from_expansion())
```
Comment (estebank) [inline src/librustc_lint/unused.rs]: Isn't this the same as `!value.span.from_expansion()`?

**Comment by estebank** [comment]
Based on the discussion in https://github.com/rust-lang/rust/issues/66295, r=me once my nitpick is addressed or answered.

**Comment by nikomatsakis** [comment]
Accepted for backport in [today's design meeting](https://rust-lang.zulipchat.com/#narrow/stream/131828-t-compiler/topic/design.20meeting.202019-12-13)

**Pull request merged**